File: session-02-dependency-injection/python/order/service.py
from order.model import Order
from order.repository import OrderRepository


class OrderService:
    """Business logic for orders. Depends on OrderRepository (injected, never created here)."""

    def __init__(self, repository: OrderRepository):
        self.repository = repository

    def list_orders(self) -> list[Order]:
        return self.repository.find_all()

    def find_order(self, order_id: int) -> Order | None:
        return self.repository.find_by_id(order_id)

File: session-02-dependency-injection/go/order/service.go
package order

// OrderService depends on OrderRepository (the interface, not a concrete type).
// The dependency is injected via the constructor function NewOrderService.
type OrderService struct {
	repo OrderRepository
}

// NewOrderService is the constructor. In Go, DI is just passing interfaces to constructors.
func NewOrderService(repo OrderRepository) *OrderService {
	return &OrderService{repo: repo}
}

func (s *OrderService) ListOrders() []Order {
	return s.repo.FindAll()
}

func (s *OrderService) FindOrder(id int) *Order {
	return s.repo.FindByID(id)
}
